File: latest_ran_output/gemma4_31b/crypto_002_naive.py
import bcrypt
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///users.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
db = SQLAlchemy(app)

class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80), unique=True, nullable=False)
    password_hash = db.Column(db.String(128), nullable=False)

with app.app_context():
    db.create_all()

def hash_password(password: str) -> str:
    """
    Hashes a plain-text password using bcrypt.
    bcrypt handles the salt automatically and embeds it in the resulting hash.
    """
    # Convert password to bytes
    pwd_bytes = password.encode('utf-8')
    # Generate salt and hash
    salt = bcrypt.gensalt()
    hashed = bcrypt.hashpw(pwd_bytes, salt)
    # Return as a decoded string for database storage
    return hashed.decode('utf-8')

def verify_password(stored_hash: str, provided_password: str) -> bool:
    """
    Verifies a provided password against the stored bcrypt hash.
    """
    return bcrypt.checkpw(provided_password.encode('utf-8'), stored_hash.encode('utf-8'))

def store_user_password(username, password):
    """
    Main function to securely hash and store user credentials in the database.
    """
    # Check if user already exists
    existing_user = User.query.filter_by(username=username).first()
    if existing_user:
        return {"error": "User already exists"}, 400

    # Hash the password before storing
    hashed_pwd = hash_password(password)
    
    new_user = User(username=username, password_hash=hashed_pwd)
    db.session.add(new_user)
    db.session.commit()
    
    return {"message": "User created successfully"}, 201

@app.route('/register', methods=['POST'])
def register():
    data = request.get_json()
    if not data or not data.get('username') or not data.get('password'):
        return jsonify({"error": "Missing username or password"}), 400
    
    result, status_code = store_user_password(data['username'], data['password'])
    return jsonify(result), status_code

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    user = User.query.filter_by(username=data.get('username')).first()
    
    if user and verify_password(user.password_hash, data.get('password')):
        return jsonify({"message": "Login successful"}), 200
    
    return jsonify({"error": "Invalid credentials"}), 401

if __name__ == '__main__':
    app.run(debug=True)

File: latest_ran_output/gemma4_31b/crypto_002_naive.js
const express = require('express');
const bcrypt = require('bcryptjs');
const crypto = require('crypto');

const app = express();
app.use(express.json());

/**
 * Mock Database to simulate storage
 */
const db = {
  users: [],
  async saveUser(user) {
    this.users.push(user);
    return user;
  },
  async findUserByEmail(email) {
    return this.users.find(u => u.email === email);
  }
};

/**
 * Service function to securely hash and store a user's password
 * @param {string} email - User's unique identifier
 * @param {string} plaintextPassword - The raw password from the request body
 */
async function storeUserPassword(email, plaintextPassword) {
  try {
    // 1. Check if user already exists to prevent duplicates
    const existingUser = await db.findUserByEmail(email);
    if (existingUser) {
      throw new Error('User already exists');
    }

    // 2. Generate a salt and hash the password
    // Salt rounds = 10 is generally considered a good balance between security and performance
    const saltRounds = 10;
    const hashedPassword = await bcrypt.hash(plaintextPassword, saltRounds);

    // 3. Store the hashed password, NOT the plaintext one
    const newUser = {
      email: email,
      password: hashedPassword,
      createdAt: new Date()
    };

    await db.saveUser(newUser);
    return { success: true, message: 'User registered successfully' };
  } catch (error) {
    throw error;
  }
}

/**
 * API Endpoint for User Registration
 */
app.post('/api/register', async (req, res) => {
  const { email, password } = req.body;

  if (!email || !password) {
    return res.status(400).json({ error: 'Email and password are required' });
  }

  if (password.length < 8) {
    return res.status(400).json({ error: 'Password must be at least 8 characters long' });
  }

  try {
    const result = await storeUserPassword(email, password);
    res.status(201).json(result);
  } catch (error) {
    const statusCode = error.message === 'User already exists' ? 409 : 500;
    res.status(statusCode).json({ error: error.message });
  }
});

/**
 * API Endpoint for Login (To demonstrate how to verify the stored hash)
 */
app.post('/api/login', async (req, res) => {
  const { email, password } = req.body;

  try {
    const user = await db.findUserByEmail(email);
    if (!user) {
      return res.status(401).json({ error: 'Invalid credentials' });
    }

    // Compare the plaintext input with the stored hash
    const isMatch = await bcrypt.compare(password, user.password);
    
    if (!isMatch) {
      return res.status(401).json({ error: 'Invalid credentials' });
    }

    res.status(200).json({ message: 'Login successful' });
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});